[WebGL] Allocate PixelBuffer for read back in web process

mwyrzykowski opened the pull request:
#### 1404a800df889699d23e23b63a1d018aeb3c1f2e
<pre>
[WebGL] Allocate PixelBuffer for read back in web process
<a href="https://bugs.webkit.org/show_bug.cgi?id=275761">https://bugs.webkit.org/show_bug.cgi?id=275761</a>
&lt;radar://129677619&gt;

Reviewed by NOBODY (OOPS!).

Allocate PixelBuffer in web process and pass through to GPU process
to avoid memory allocation in the GPU process.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/PixelBuffer.h:
(WebCore::PixelBuffer::isShareablePixelBuffer const):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::readPixelsForPaintResults):
(WebCore::GraphicsContextGLANGLE::updatePixelsForPaintResults):
(WebCore::GraphicsContextGLANGLE::drawingBufferToPixelBuffer):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebKit/GPUProcess/graphics/ImageBufferShareableAllocator.cpp:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLCocoa.cpp:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(drawingBufferToPixelBuffer):
* Source/WebKit/GPUProcess/graphics/ShareablePixelBuffer.cpp: Removed.
* Source/WebKit/GPUProcess/graphics/ShareablePixelBuffer.h: Removed.
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.cpp:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.h:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::drawingBufferToPixelBuffer):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1404a800df889699d23e23b63a1d018aeb3c1f2e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56664 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35991 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9137 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60278 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7105 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58791 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43614 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7296 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45899 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4981 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58694 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33832 "Found 9 new test failures: fast/canvas/webgl/canvas-drawImage.html fast/canvas/webgl/canvas-getImageData.html fast/canvas/webgl/premultiplyalpha-test.html fast/css-grid-layout/grid-container-percentage-rows.html http/wpt/webcodecs/videoFrame-webglCanvasImageSource.html webgl/1.0.x/conformance/canvas/to-data-url-test.html webgl/1.0.x/conformance/context/premultiplyalpha-test.html webgl/2.0.y/conformance/canvas/to-data-url-test.html webgl/2.0.y/conformance/context/premultiplyalpha-test.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48909 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26758 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/56190 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30611 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6240 "Found 4 new test failures: fast/canvas/webgl/canvas-drawImage.html fast/canvas/webgl/canvas-getImageData.html fast/canvas/webgl/premultiplyalpha-test.html imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/dialog-autofocus.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6108 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52557 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6512 "Found 8 new test failures: fast/canvas/webgl/canvas-drawImage.html fast/canvas/webgl/canvas-getImageData.html fast/canvas/webgl/premultiplyalpha-test.html http/wpt/webcodecs/videoFrame-webglCanvasImageSource.html webgl/1.0.x/conformance/canvas/to-data-url-test.html webgl/1.0.x/conformance/context/premultiplyalpha-test.html webgl/2.0.y/conformance/canvas/to-data-url-test.html webgl/2.0.y/conformance/context/premultiplyalpha-test.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61961 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/576 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6609 "Found 8 new test failures: fast/canvas/webgl/canvas-drawImage.html fast/canvas/webgl/canvas-getImageData.html fast/canvas/webgl/premultiplyalpha-test.html http/wpt/webcodecs/videoFrame-webglCanvasImageSource.html webgl/1.0.x/conformance/canvas/to-data-url-test.html webgl/1.0.x/conformance/context/premultiplyalpha-test.html webgl/2.0.y/conformance/canvas/to-data-url-test.html webgl/2.0.y/conformance/context/premultiplyalpha-test.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53157 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/578 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48967 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53172 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/496 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31819 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32905 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33989 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32651 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/html/HTMLCanvasElement.cpp:
```diff
@@ -780,7 +780,8 @@ RefPtr<ImageData> HTMLCanvasElement::getImageData()
         ResourceLoadObserver::shared().logCanvasRead(document());
 
-    RefPtr pixelBuffer = dynamicDowncast<ByteArrayPixelBuffer>(context->drawingBufferToPixelBuffer(GraphicsContextGL::FlipY::Yes));
+    RefPtr<ByteArrayPixelBuffer> pixelBuffer { };
```
This isn't allocating anything, and you don't need to change this signature. The only thing needed would be allocation at the IPC level.

kkinnunen-apple commented on Source/WebCore/platform/graphics/GraphicsContextGL.h:
```diff
@@ -1654,7 +1654,7 @@ class GraphicsContextGL : public RefCounted<GraphicsContextGL> {
     virtual RefPtr<VideoFrame> surfaceBufferToVideoFrame(SurfaceBuffer) = 0;
 #endif
-    virtual RefPtr<PixelBuffer> drawingBufferToPixelBuffer(FlipY) = 0;
+    virtual RefPtr<PixelBuffer> drawingBufferToPixelBuffer(FlipY, RefPtr<WebCore::PixelBuffer>&&) = 0;
```
I don't think these sort of APIs make sense to add, e.g. accepts pixelbuffer and returns pixelbuffer.

Currently polymorphic creation of PixelBuffer is not possible inside the API, so it creates a complication.
Same complication exists in the ImageBuffer case. See below.

The main API should likely be as it is today, and then the IPC implemetation should likely follow same pattern as

surfaceBufferToImageBuffer -> drawSurfaceBufferToImageBuffer

drawingBufferToPixelBuffer -> drawDrawingBufferToImageBuffer.

mwyrzykowski committed:
[WebGL] Allocate PixelBuffer for read back in web process
https://bugs.webkit.org/show_bug.cgi?id=275761
<radar://129677619>

Reviewed by NOBODY (OOPS!).

Allocate PixelBuffer in web process and pass through to GPU process
to avoid memory allocation in the GPU process.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/PixelBuffer.h:
(WebCore::PixelBuffer::isShareablePixelBuffer const):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::readPixelsForPaintResults):
(WebCore::GraphicsContextGLANGLE::updatePixelsForPaintResults):
(WebCore::GraphicsContextGLANGLE::drawingBufferToPixelBuffer):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebKit/GPUProcess/graphics/ImageBufferShareableAllocator.cpp:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLCocoa.cpp:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(drawingBufferToPixelBuffer):
* Source/WebKit/GPUProcess/graphics/ShareablePixelBuffer.cpp: Removed.
* Source/WebKit/GPUProcess/graphics/ShareablePixelBuffer.h: Removed.
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.cpp:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.h:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::drawingBufferToPixelBuffer):

kkinnunen-apple commented on Source/WebCore/platform/graphics/ShareablePixelBuffer.cpp:
```diff
@@ -26,10 +26,9 @@
 #include "ShareablePixelBuffer.h"
 
-#include <WebCore/SharedMemory.h>
+#include "SharedMemory.h"
```
This stuff is Source/WebKit stuff, shouldn't exist in WebCore unless absolutely neccesary

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
```diff
@@ -3047,17 +3048,25 @@ void RemoteGraphicsContextGLProxy::setDrawingBufferColorSpace(const WebCore::Des
 
+void RemoteGraphicsContextGLProxy::drawingBufferToPixelBuffer(WebCore::GraphicsContextGLFlipY, RefPtr<WebCore::PixelBuffer>&&)
+{
+    RELEASE_ASSERT_NOT_REACHED();
```
This is a symptom of doing something incorrect. 

E.g. interface has two functions and two backends. One of the functions is invalid for each backend.

kkinnunen-apple replied:
Also, the ffile is Generated, you cannot go around modifying it by hand.

kkinnunen-apple commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -257,6 +257,16 @@ RefPtr<PixelBuffer> GraphicsContextGLANGLE::readPixelsForPaintResults()
+
+void GraphicsContextGLANGLE::updatePixelsForPaintResults(RefPtr<PixelBuffer>&& pixelBuffer)
+{
+    if (pixelBuffer->size() != getInternalFramebufferSize())
```
This is GPUP process validation, it should not be here.
And it is not sufficient, as you need to validate all the other expected propertioes (pixelformat, shm size)

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
```diff
@@ -3047,17 +3048,25 @@ void RemoteGraphicsContextGLProxy::setDrawingBufferColorSpace(const WebCore::Des
         return { };
-    auto sendResult = sendSync(Messages::RemoteGraphicsContextGL::DrawingBufferToPixelBuffer(arg0));
+    PixelBufferFormat format { AlphaPremultiplication::Unpremultiplied, PixelFormat::RGBA8, DestinationColorSpace::SRGB() };
+    auto pixelBuffer = ShareablePixelBuffer::tryCreate(format, WebCore::IntSize(m_currentWidth, m_currentHeight));
```
The caller actually needs a ByteArrayPixelBuffer. This is due to the way WebCore::ImageData works.

I think this patch produces just black, at least in most cases..
